### PR TITLE
Add expand composite rule

### PR DESCRIPTION
**Pull request by xysheng-baidu**
### PR types

Others
### PR changes

Others
### Describe

add composite rule for expand

### CODE REVIEW

File: python/paddle/incubate/autograd/composite_rules.py [hunk 1]
```diff
@@ -192,6 +192,42 @@ def mean_composite(x, axis, keepdim):
...
+def expand_v2_composite(x, shape):
+    """
+    define composite rule of op expnad_v2, expand_v2->expand
+    repeat_times = shape / x.shape
+    out = tile(x, repeat_times = repeat_times)
+    """
+    shape_in = x.shape
+    assert len(shape) >= len(shape_in)
+    dim_out = len(shape)
+    dim_in = len(shape_in)
```
Comment (cyber-pioneer) [inline python/paddle/incubate/autograd/composite_rules.py:204]: polish 3 lines

File: python/paddle/incubate/autograd/composite_rules.py [hunk 2]
```diff
@@ -192,6 +192,42 @@ def mean_composite(x, axis, keepdim):
...
+    if dim_out == 0:
+        return x
```
Comment (cyber-pioneer) [inline python/paddle/incubate/autograd/composite_rules.py]: image

test case does not cover this branch, add relative case

Comment (xysheng-baidu) [reply python/paddle/incubate/autograd/composite_rules.py]: dim_out can't be 0, remove the code.

File: python/paddle/fluid/tests/unittests/test_expand_v2_op.py [hunk 3]
```diff
@@ -332,6 +332,136 @@ def test_grad(self):
...
+        self.attrs = {'shape': [2, 4, 5]}
+        output = np.tile(self.inputs['X'], (1, 1, 1))
+        self.outputs = {'Out': output}
+
+    def test_check_output(self):
+        self.check_output()
+
+
+#  Situation 10: comp case, input x is Bool
+class TestExpandV2CompOpBoolean(OpTest):
```
Comment (Charles-hit) [inline python/paddle/fluid/tests/unittests/test_expand_v2_op.py:406]: 之前没有组合测试的单测可以去掉，例如Shape Tensor类的

**Review by cyber-pioneer — APPROVED** [review]

File: python/paddle/incubate/autograd/composite_rules.py [hunk 4]
```diff
@@ -192,6 +192,40 @@ def mean_composite(x, axis, keepdim):
...
+        else:
+            assert size_out % size_in == 0
+            repeat = int(size_out / size_in)
+        repeat_times.append(repeat)
+    if dim_in < dim_out:
+        shape_in_expand = []
+        for i in range(dim_out - dim_in):
+            shape_in_expand.append(1)
+        shape_in_expand.extend(shape_in)
+        x_reshape = reshape(x, shape_in_expand)
```
Comment (JiabinYang) [inline python/paddle/incubate/autograd/composite_rules.py:224]: Add comment to show why we need reshape first

**Review by Charles-hit — APPROVED** [review]

File: python/paddle/incubate/autograd/composite_rules.py (hunk 4 above)
Comment (xysheng-baidu) [reply python/paddle/incubate/autograd/composite_rules.py:224]: Under the static graph, the tile op will set the expanded dimension to -1 when expanding the tensor, which will result in failure to pass the shape check. So for tensors that need to expand the dimension, reshape will be used in advance

Comment (xysheng-baidu) [reply python/paddle/incubate/autograd/composite_rules.py:224]: @JiabinYang https://github.com/PaddlePaddle/Paddle/pull/51564

**Pull request merged**